### PR TITLE
Fix query expression parsing and recovery issues

### LINKED ISSUE #25316: Bad Sad with "java.lang.IllegalStateException: found close-brace in: var-decl-stmt"
**Steps to reproduce:**
```
import ballerina/io;

public function main() {
    stream<Report> reportStream = stream from var student in studentList
       where student.gpa >= 2.0
       let string degreeName = "Bachelor of Medicine",
       int graduationYear = calGraduationYear(student.intakeYear)
       select {
              name: student.firstName + " " + student.lastName,
              degree: degreeName,
              graduationYear: graduationYear
       }
       limit 2;

    json j = from var r in reportStream select r.toJson()
    io:println(j.toJsonString());
}
```
[ballerina-internal.log](https://github.com/ballerina-platform/ballerina-lang/files/5093044/ballerina-internal.log)

**Affected Versions:**
Swan Lake Preview 3

### LINKED ISSUE #25542: Fix error recovery for `equals` operator
**Description:**
$title and implement explicit recovery rules for `equals`. Also, restrict that to `join on` context.

**Steps to reproduce:**
n/a

**Affected Versions:**
n/a

**OS, DB, other environment details and versions:**
n/a

**Related Issues/PRs (optional):**
https://github.com/ballerina-platform/ballerina-lang/pull/25473

### LINKED ISSUE #25590: Query expression context has not been ended properly in BallerinaParser
**Description:**
Consider the following code snippet

```bal
function foo() {
    var query = from var a in b select c;
}

type bar Object {
```
It throws the following exception
```
Exception in thread "main" java.lang.IllegalStateException: found close-brace in: func-def
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRuleForCloseBrace(BallerinaParserErrorHandler.java:3595)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.getNextRule(BallerinaParserErrorHandler.java:2171)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:387)
	at io.ballerinalang.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:329)
	at io.ballerinalang.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1053)
```
**Suggested Labels:**
Type/Bug, Priority/Blocker, Area/Language, Compiler/Parser

**Suggested Assignees (optional):**
@lochana-chathura

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
* Fix `query-expression` context ending in the BallerinaParser when there's no `on-conflict-clause`
* Remove parser support of `equals` as an operator
* Fix operator precedence param of the join-on-condition expression parsing
* Implement explicit recovery rules for `equals` keyword of `join-on-condition` in BallerinaParserErrorHandler

Fixes #25542 
Fixes #25590 
Fixes #25316

## Approach
N/A

## Samples
N/A

## Remarks
N/A

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -9504,17 +9501,35 @@ private STNode parseJoinClause(boolean isRhsExpr) {
...
      * Parse on clause.
      * <p>
-     * <code>on clause := on expression</code>
+     * <code>on clause := `on` expression `equals` expression</code>
      *
      * @return On clause node
      */
     private STNode parseOnClause(boolean isRhsExpr) {
+        STToken nextToken = peek();
+        if (isQueryClauseStartToken(nextToken.kind)) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9510]: Don't we need to recover here?
What if theres an additional token (say a `public` keyword) before the `on` keyword?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9510]: shouldn't that extra token be recovered as invalid in parseOnKeyword() beneath?

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9510]: ah this is inverted. Sorry, my question should be inverted too :) 
What if theres an additional token before an actual query-clause starting token?
e.g:
```ballerina
from var person in personList
outer join Department dept in deptList
// missing on clause
someExtraToken select {
    fname : person.fname,
    lname : person.lname,
    dept : dept.name
};
```
will it keep the 'select' query as is?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9510]:
```
var a = from var person in personList
outer join Department dept in deptList
 MISSING[+]// missing on clause
someExtraToken  MISSING[on] MISSING[] MISSING[equals] MISSING[]select {
    fname : person.fname,
    lname : person.lname,
    dept : dept.name
};
```
Above is the recovery if that extra token is an identifier.  if it is `public`recovery is pretty bad yet it is a valid syntax. 
```
var a = from var person in personList
outer join Department dept in deptList
 MISSING[on] MISSING[<-]// missing on clause
 INVALID[public]  INVALID[select] {
    fname : person MISSING[,] INVALID[.]fname,
    lname : person MISSING[,] INVALID[.]lname,
    dept : dept MISSING[}].name
 MISSING[equals] MISSING[] MISSING[select] MISSING[] INVALID[}];
```
If we are going to handle when there's a  non-identifier token before the beginning of next clause in absence of join-on-condition it is like we are making join-on-condition optional in the error handler. As we don't allow that non-identifier to be part of join-on-condition. I think if there are extra tokens they should be allowed to be a part of join-on-condition recovery.
WDYT?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9510]: We can do something like this in the error handler though. before parsing on-clause as next rule, we check if the next lookahead token is one of query starting tokens and if and only if, we skip join-on-condition path in the error handler. I feel that way recovery logic is consistence with parsing logic.

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9510]: hmm, I feel which ever option we take, there will be consequences.
We can go ahead with the current approach, and can revisit if it becomes an actual issue.

**Review by SupunS — APPROVED** [review]

**Pull request merged**